Add additional WKWebView SPI to extract rendered text for a given CSS selector in the main frame

whsieh opened the pull request:
#### 503f7e1ff8cef2d718e4dfa3612e1d9907d81e67
<pre>
Add additional WKWebView SPI to extract rendered text for a given CSS selector in the main frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=270544">https://bugs.webkit.org/show_bug.cgi?id=270544</a>

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Add another `WKWebView` SPI method to extract rendered text from an element in the main frame, given
a CSS selector. This flattens all text content in the element subtree (including subframes) into a
single string, sorted in block layout direction order by the root view bounds of each rendered text
node.

Test: fast/text-extraction/extract-rendered-text-with-subframe.html

* LayoutTests/fast/text-extraction/extract-rendered-text-with-subframe-expected.txt: Added.
* LayoutTests/fast/text-extraction/extract-rendered-text-with-subframe.html: Added.

Add a layout test to exercise this SPI.

* LayoutTests/fast/text-extraction/resources/subframe.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.requestTextExtraction):
(window.UIHelper.requestRenderedTextForSelector):

Add a testing hook to exercise the new SPI from layout tests.

(window.UIHelper):
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractRenderedText):

Implement the main logic here, by recursively traversing through the render tree to accumulate text
in a `Vector` of offsets (based on block writing direction — e.g. y-offsets, in top-down LTR text)
and strings, filtered to only contain letters (including non-Latin script) and numbers. We also
represent replaced renderers with `{w,h}`, representing the size of the renderer in root view
coordinates.

The final output is a space-separated string of extracted tokens.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestRenderedTextForElementSelector:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestTextExtraction):
(WebKit::WebPageProxy::requestRenderedTextForElementSelector):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestRenderedTextForElementSelector):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::requestRenderedTextForSelector):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestRenderedTextForSelector):

Canonical link: <a href="https://commits.webkit.org/275728@main">https://commits.webkit.org/275728@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77220c540eb16aa99b7259a4d045348f05353f07

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42636 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21658 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45038 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45245 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38760 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44943 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/25320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19023 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35296 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43210 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/25320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36697 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16246 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/25320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37756 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/693 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/25320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38083 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46739 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/17455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14381 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41994 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37025 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40613 "Found 2 new API test failures: /WebKitGTK/TestWebKitFaviconDatabase:/webkit/WebKitFaviconDatabase/get-favicon, /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9530 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19253 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18719 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -424,5 +433,137 @@ Item extractItem(std::optional<WebCore::FloatRect>&& collectionRectInRootView, P
+        return;
+
+    static constexpr auto minOpacityToConsiderVisible = 0.05;
+    static constexpr auto epsilonForNearbyText = 5;
```
Is this a pixel count?

aprotyas replied:
also nit: you can move this inside `appendStrings` since it isn't used outside that scope

whsieh replied:
> Is this a pixel count?

Yes — I'll clarify this by renaming it to `minPixelDistanceForNearbyText`

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2761,6 +2763,19 @@ - (void)_addAppHighlightInNewGroup:(BOOL)newGroup originatedInApp:(BOOL)originat
+- (void)_requestRenderedTextForElementSelector:(NSString *)selector completionHandler:(void(^)(NSString *, NSError *))completion
+{
+    _page->requestRenderedTextForElementSelector(selector, [completion = makeBlockPtr(completion)](Expected<String, WebCore::ExceptionCode>&& result) {
+        if (result.has_value())
```
I think just `if (result)` should work?

pxlcoder commented on Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
```diff
@@ -336,4 +336,16 @@ - (void)paste:(id)sender;
+        if (!m_context)
+            return;
+
+        auto result = adopt(JSStringCreateWithCFString((__bridge CFStringRef)(error.description ?: text)));
```
I wonder if we should use `JSRetainPtr` rather than `auto` due to the explicit smart pointer usage guidelines.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9345,6 +9346,15 @@ void WebPage::requestTextExtraction(std::optional<FloatRect>&& collectionRectInR
 
+void WebPage::requestRenderedTextForElementSelector(String&& selector, CompletionHandler<void(Expected<String, WebCore::ExceptionCode>&&)>&& completion)
+{
+    auto mainFrame = dynamicDowncast<LocalFrame>(corePage()->protectedMainFrame());
```
`RefPtr`?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2761,6 +2763,19 @@ - (void)_addAppHighlightInNewGroup:(BOOL)newGroup originatedInApp:(BOOL)originat
+        if (result.has_value())
+            return completion(result.value(), nil);
+
+        auto exceptionName = WebCore::DOMException::name(result.error()).createNSString();
```
`auto` -> `RetainPtr`?

whsieh committed:
Add additional WKWebView SPI to extract rendered text for a given CSS selector in the main frame
https://bugs.webkit.org/show_bug.cgi?id=270544

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Add another `WKWebView` SPI method to extract rendered text from an element in the main frame, given
a CSS selector. This flattens all text content in the element subtree (including subframes) into a
single string, sorted in block layout direction order by the root view bounds of each rendered text
node.

Test: fast/text-extraction/extract-rendered-text-with-subframe.html

* LayoutTests/fast/text-extraction/extract-rendered-text-with-subframe-expected.txt: Added.
* LayoutTests/fast/text-extraction/extract-rendered-text-with-subframe.html: Added.

Add a layout test to exercise this SPI.

* LayoutTests/fast/text-extraction/resources/subframe.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.requestTextExtraction):
(window.UIHelper.requestRenderedTextForSelector):

Add a testing hook to exercise the new SPI from layout tests.

(window.UIHelper):
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractRenderedText):

Implement the main logic here, by recursively traversing through the render tree to accumulate text
in a `Vector` of offsets (based on block writing direction — e.g. y-offsets, in top-down LTR text)
and strings, filtered to only contain letters (including non-Latin script) and numbers. We also
represent replaced renderers with `{w,h}`, representing the size of the renderer in root view
coordinates.

The final output is a space-separated string of extracted tokens.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestRenderedTextForElementSelector:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestTextExtraction):
(WebKit::WebPageProxy::requestRenderedTextForElementSelector):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestRenderedTextForElementSelector):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::requestRenderedTextForSelector):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestRenderedTextForSelector):

Canonical link: https://commits.webkit.org/275728@main
